packages/web-components/fast-components/src/design-system-provider/design-system-provider.ts

```diff
@@ -71,16 +70,69 @@ export function isDesignSystemProvider(
     return (el as any).isDesignSystemProvider;
 }
 
-/**
- * Slight hack to get intelisense and type checking for the
- * @designSystemConsumer decorator because TypeScript does
- * not allow type mutation for decorators.
- * https://github.com/microsoft/TypeScript/issues/4881
- */
-/* eslint-disable-next-line @typescript-eslint/no-empty-interface */
-export interface DesignSystemProvider extends DesignSystemConsumer {}
-@designSystemConsumer
-export class DesignSystemProvider extends FastElement {
+export class DesignSystemProvider extends FastElement
+    implements CSSCustomPropertyTarget, DesignSystemConsumer {
+    private customPropertyBehaviors: Map<
+        string,
+        CSSCustomPropertyDefinition & { count: number }
+    > = new Map();
+
+    public registerCSSCustomProperty(behavior: CSSCustomPropertyDefinition) {
+        const cached = this.customPropertyBehaviors.get(behavior.name);
+
+        if (cached) {
+            cached.count += 1;
+        } else {
+            this.customPropertyBehaviors.set(behavior.name, { ...behavior, count: 1 });
+            this.writeCustomProperty(behavior);
+        }
+    }
+
+    public unregisterCSSCustomProperty(behavior: CSSCustomPropertyDefinition) {
+        const cached = this.customPropertyBehaviors.get(behavior.name);
+
+        if (cached) {
+            cached.count -= 1;
+
+            if (cached.count === 0) {
+                this.customPropertyBehaviors.delete(behavior.name);
+                this.customPropertyManager.delete(behavior);
+            }
+        }
+    }
+
+    @observable
+    public provider: DesignSystemProvider | null = null;
+    private providerChanged(
+        prev: DesignSystemProvider | null,
+        next: DesignSystemProvider | null
+    ): void {
+        if (prev instanceof HTMLElement) {
+            Object.keys(prev.designSystemProperties).forEach(key => {
+                Observable.getNotifier(prev.designSystem).unsubscribe(
+                    this.providerDesignSystemChangeHandler,
+                    key
+                );
+            });
+        }
+
+        if (next instanceof HTMLElement && isDesignSystemProvider(next)) {
+            Object.keys(next.designSystemProperties).forEach(key => {
+                Observable.getNotifier(next.designSystem).subscribe(
+                    this.providerDesignSystemChangeHandler,
+                    key
+                );
+            });
+
+            this.syncDesignSystemWithProvider();
+        }
+    }
+
+    constructor() {
+        super();
+
+        this.$fastController.addBehaviors([new DesignSystemConsumerBehavior()]);
```

Comment:
@EisenbergEffect I want to make sure I wasn't circumventing anything here by adding behaviors to the controller - does this look right?

Reply:
Looks good. I tried to design that so that arbitrary behaviors could be added independent of the style association part.